File: test/src/memory_test_harness.ts
#!/usr/bin/env node
/**
 * Memory Test Harness for Socratic Shell
 * 
 * Tests Claude Code interaction patterns for memory bank implementation.
 * Uses node-pty for reliable terminal interaction.
 */

import * as pty from 'node-pty';

interface TestResult {
    success: boolean;
    found: string[];
    missing?: string[];
    responseLength: number;
}

interface BootTestResult extends TestResult {
    missing: string[];
}

interface ConsolidationTestResult extends TestResult {}

class ClaudeTestHarness {
    private ptyProcess: pty.IPty | null = null;
    private output: string = '';
    private isReady: boolean = false;

    async start(): Promise<boolean> {
        console.log('🚀 Starting Claude Code with node-pty...');
        
        // Spawn Claude Code with proper environment
        this.ptyProcess = pty.spawn('/home/nikomatsakis/.claude/local/claude', [], {
            name: 'xterm-256color',
            cols: 120,
            rows: 40,
            cwd: process.cwd(),
            env: process.env
        });

        // Set up data handlers
        this.ptyProcess.onData((data: string) => {
            this.output += data;
            // Look for readiness indicators
            if (data.includes('>') || data.includes('claude')) {
                this.isReady = true;
            }
        });

        this.ptyProcess.onExit(({ exitCode, signal }) => {
            console.log(`Claude exited with code ${exitCode}, signal ${signal}`);
        });

        // Wait for Claude to be ready
        console.log('⏳ Waiting for Claude to be ready...');
        await this.waitForReady();
        console.log('✅ Claude is ready!');
        
        return true;
    }

    private async waitForReady(timeout: number = 30000): Promise<void> {
        const start = Date.now();
        while (!this.isReady && (Date.now() - start) < timeout) {
            await new Promise(resolve => setTimeout(resolve, 100));
        }
        if (!this.isReady) {
            throw new Error('Timeout waiting for Claude to be ready');
        }
    }

    async sendMessage(message: string): Promise<string> {
        if (!this.ptyProcess) {
            throw new Error('Claude process not started');
        }

        console.log(`📤 Sending: ${message}`);
        
        // Clear output buffer
        this.output = '';
        
        // Send message and submit with Enter keys
        this.ptyProcess.write(message);
        await new Promise(resolve => setTimeout(resolve, 100));
        this.ptyProcess.write('\r');
        await new Promise(resolve => setTimeout(resolve, 100));
        this.ptyProcess.write('\x0D');
        
        // Wait for response
        await this.waitForResponse();
        
        // Clean up the output
        const cleanOutput = this.stripAnsi(this.output);
        console.log(`📨 Response length: ${cleanOutput.length} characters`);
        
        return cleanOutput;
    }

    private async waitForResponse(timeout: number = 60000): Promise<void> {
        const start = Date.now();
        let stableCount = 0;
        let lastLength = 0;
        
        console.log('⏳ Waiting for response...');
        
        while ((Date.now() - start) < timeout) {
            await new Promise(resolve => setTimeout(resolve, 500));
            
            // Check if output has stabilized
            if (this.output.length === lastLength) {
                stableCount++;
                if (stableCount >= 6) { // Stable for 3 seconds
                    console.log('✅ Output appears stable');
                    break;
                }
            } else {
                stableCount = 0;
                lastLength = this.output.length;
            }
        }
        
        if ((Date.now() - start) >= timeout) {
            console.log('⏰ Timeout reached');
        }
    }

    private stripAnsi(text: string): string {
        // Remove ANSI escape sequences
        return text.replace(/\x1b\[[0-9;]*[mGKHJA-Z]/g, '');
    }

    close(): void {
        if (this.ptyProcess) {
            this.ptyProcess.kill();
        }
    }
}

/**
 * Test scenarios for memory bank implementation
 */
class MemoryBankTests {
    constructor(private harness: ClaudeTestHarness) {}

    async testBootProcedure(): Promise<BootTestResult> {
        console.log('\n=== Memory Test: Boot Procedure ===');
        const response = await this.harness.sendMessage('Hi again, Claude!');
        
        const indicators = [
            'Prime Directive',
            'Make it so', 
            'partnership',
            'hooks',
            'completion'
        ] as const;
        
        console.log('🔍 Checking for boot procedure patterns:');
        const found: string[] = [];
        for (const indicator of indicators) {
            if (response.toLowerCase().includes(indicator.toLowerCase())) {
                console.log(`🎉 FOUND: ${indicator}`);
                found.push(indicator);
            } else {
                console.log(`❌ Missing: ${indicator}`);
            }
        }
        
        const missing = indicators.filter(i => !found.includes(i));
        
        return {
            success: found.length === indicators.length,
            found,
            missing,
            responseLength: response.length
        };
    }

    async testMemoryConsolidationTrigger(): Promise<ConsolidationTestResult> {
        console.log('\n=== Memory Test: Consolidation Trigger ===');
        
        // Simulate a complex task that should trigger consolidation
        const response = await this.harness.sendMessage('I need to implement a complex feature with multiple steps. Can you help me plan it out?');
        
        // Look for consolidation/planning patterns
        const consolidationIndicators = [
            'todo',
            'plan',
            'steps',
            'track',
            'organize'
        ] as const;
        
        console.log('🔍 Checking for consolidation triggers:');
        const found: string[] = [];
        for (const indicator of consolidationIndicators) {
            if (response.toLowerCase().includes(indicator.toLowerCase())) {
                console.log(`🎯 FOUND: ${indicator}`);
                found.push(indicator);
            }
        }
        
        return {
            success: found.length > 0,
            found,
            responseLength: response.length
        };
    }
}

async function main(): Promise<void> {
    const harness = new ClaudeTestHarness();
    const tests = new MemoryBankTests(harness);
    
    try {
        await harness.start();
        
        // Run memory bank tests
        const bootResult = await tests.testBootProcedure();
        const consolidationResult = await tests.testMemoryConsolidationTrigger();
        
        // Summary
        console.log('\n=== Test Summary ===');
        console.log(`Boot Procedure: ${bootResult.success ? '✅ PASS' : '❌ FAIL'}`);
        console.log(`Consolidation Trigger: ${consolidationResult.success ? '✅ PASS' : '❌ FAIL'}`);
        
        if (!bootResult.success) {
            console.log(`Boot missing: ${bootResult.missing.join(', ')}`);
        }
        
        console.log('🎉 Memory test harness working successfully!');
        
    } catch (error) {
        console.error('❌ Error:', error);
    } finally {
        harness.close();
    }
}

if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}

File: dialectic/memory_test_harness.py
#!/usr/bin/env python3
"""
Memory Test Harness for Socratic Shell

Tests Claude Code interaction patterns for memory bank implementation.
Uses the claude-code SDK for clean, reliable API interaction.
"""

import anyio
from dataclasses import dataclass
from typing import List

from claude_code_sdk import query, AssistantMessage, TextBlock


@dataclass
class TestResult:
    success: bool
    found: List[str]
    missing: List[str]
    response_length: int
    response_preview: str


class MemoryBankTests:
    """Test scenarios for memory bank implementation"""

    async def test_boot_procedure(self) -> TestResult:
        """Test Claude's boot procedure with collaboration patterns"""
        print('\n=== Memory Test: Boot Procedure ===')
        
        response_text = ""
        async for message in query(prompt='Hi again, Claude!'):
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        response_text += block.text
        
        indicators = [
            'Prime Directive',
            'Make it so', 
            'partnership',
            'hooks',
            'completion'
        ]
        
        print('🔍 Checking for boot procedure patterns:')
        found = []
        for indicator in indicators:
            if indicator.lower() in response_text.lower():
                print(f'🎉 FOUND: {indicator}')
                found.append(indicator)
            else:
                print(f'❌ Missing: {indicator}')
        
        missing = [i for i in indicators if i not in found]
        
        return TestResult(
            success=len(found) == len(indicators),
            found=found,
            missing=missing,
            response_length=len(response_text),
            response_preview=response_text[-800:]  # Last 800 chars
        )

    async def test_memory_consolidation_trigger(self) -> TestResult:
        """Test Claude's response to complex tasks that should trigger consolidation"""
        print('\n=== Memory Test: Consolidation Trigger ===')
        
        response_text = ""
        async for message in query(prompt='I need to implement a complex feature with multiple steps. Can you help me plan it out?'):
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        response_text += block.text
        
        consolidation_indicators = [
            'todo',
            'plan',
            'steps',
            'track',
            'organize'
        ]
        
        print('🔍 Checking for consolidation triggers:')
        found = []
        for indicator in consolidation_indicators:
            if indicator.lower() in response_text.lower():
                print(f'🎯 FOUND: {indicator}')
                found.append(indicator)
        
        missing = [i for i in consolidation_indicators if i not in found]
        
        return TestResult(
            success=len(found) > 0,
            found=found,
            missing=missing,
            response_length=len(response_text),
            response_preview=response_text[-500:]  # Last 500 chars
        )

    async def test_memory_retrieval_trigger(self) -> TestResult:
        """Test scenarios that should trigger memory retrieval"""
        print('\n=== Memory Test: Retrieval Trigger ===')
        
        # First establish some context
        async for message in query(prompt='I use vim keybindings and prefer minimal UI.'):
            pass  # Just establish context
        
        # Then ask about something that should trigger memory retrieval
        response_text = ""
        async for message in query(prompt='What do you know about my preferences?'):
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        response_text += block.text
        
        retrieval_indicators = [
            'vim',
            'keybindings', 
            'minimal',
            'preferences',
            'remember'
        ]
        
        print('🔍 Checking for memory retrieval:')
        found = []
        for indicator in retrieval_indicators:
            if indicator.lower() in response_text.lower():
                print(f'🧠 FOUND: {indicator}')
                found.append(indicator)
        
        missing = [i for i in retrieval_indicators if i not in found]
        
        return TestResult(
            success=len(found) >= 2,  # Should remember at least 2 elements
            found=found,
            missing=missing,
            response_length=len(response_text),
            response_preview=response_text[-500:]
        )



async def main():
    """Run all memory bank tests"""
    tests = MemoryBankTests()
    
    try:
        print('🚀 Starting Memory Bank Test Suite')
        print('Using claude-code SDK for clean API interaction')
        
        # Run all tests
        boot_result = await tests.test_boot_procedure()
        consolidation_result = await tests.test_memory_consolidation_trigger()
        retrieval_result = await tests.test_memory_retrieval_trigger()
        
        # Summary
        print('\n=== Test Summary ===')
        print(f'Boot Procedure: {"✅ PASS" if boot_result.success else "❌ FAIL"}')
        print(f'Consolidation Trigger: {"✅ PASS" if consolidation_result.success else "❌ FAIL"}')
        print(f'Memory Retrieval: {"✅ PASS" if retrieval_result.success else "❌ FAIL"}')
        
        # Show missing items for failed tests
        if not boot_result.success:
            print(f'Boot missing: {", ".join(boot_result.missing)}')
        if not consolidation_result.success:
            print(f'Consolidation missing: {", ".join(consolidation_result.missing)}')
        if not retrieval_result.success:
            print(f'Retrieval missing: {", ".join(retrieval_result.missing)}')
        
        # Overall status
        all_passed = all([boot_result.success, consolidation_result.success, retrieval_result.success])
        print(f'\n🎉 Overall: {"ALL TESTS PASSED" if all_passed else "SOME TESTS FAILED"}')
        
        # Show response lengths for debugging
        print(f'\nResponse lengths:')
        print(f'  Boot: {boot_result.response_length} chars')
        print(f'  Consolidation: {consolidation_result.response_length} chars') 
        print(f'  Retrieval: {retrieval_result.response_length} chars')
        
    except Exception as error:
        print(f'❌ Error running tests: {error}')
        raise


if __name__ == '__main__':
    anyio.run(main)